### ISSUE TITLE
incompatible types when using re-exports of the same crate at different paths

### ISSUE THREAD

**Issue by dustbort** [rust-lang/rust-analyzer#13723]
I tried this code:

```rust
let item = sickle_builder::types::time::Date::from_calendar_date(
    2020,
    sickle_builder::types::time::Month::January,
    1,
)?;
let y: i32 = item.year();
let m: tiberius::time::time::Month = item.month();
let d: u8 = item.day();
let date: tiberius::time::Date = tiberius::time::time::Date::from_calendar_date(y, m, d)?;
```

I expected to see this happen: No error

Instead, this happened: 
```
error[E0308]: `?` operator has incompatible types
  --> sickle-query/src/mssql_queryable.rs:53:21
   |
53 |                     tiberius::time::time::Date::from_calendar_date(y, m, d)?
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `tiberius::time::Date`, found struct `sickle_builder::types::time::Date`
   |
   = note: `?` operator cannot convert from `sickle_builder::types::time::Date` to `tiberius::time::Date`
```

`sickle_query::types::time` and `tiberius::time` are re-exports of the `time` crate. For multiple reasons, these may be different versions with different features and _are not be assumed to be the same_. However, it seems that rust-analyzer is detecting that they are the same crate, and then "forgets" the path or assumes the wrong path on line 53. But though the path is explicitly given on both the left and right hand sides of the assignment, so there is no ambiguity.

I do not see the error when running `cargo build`.

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

**rust-analyzer version**: 0.3.1301-standalone

**rustc version**: 1.64.0 (a55dd71d5 2022-09-19)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

vscode info:
```
Version: 1.73.0
Commit: 8fa188b2b301d36553cbc9ce1b0a146ccb93351f
Date: 2022-11-03T14:04:08.642Z
Electron: 19.1.4
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Linux x64 6.0.8-1-MANJARO
Sandboxed: No
```

**Comment by bjorn3** [comment]
The error you reported is from rustc, not rust-analyzer. Do you get the same error if you run `cargo check --workspace --all-targets`?

**Comment by dustbort** [comment]
@bjorn3 No, I do not
```
$ cargo check --workspace --all-targets                                                                                                                  ✔ 
   Compiling num-traits v0.2.15
    Checking sickle-builder v0.1.0 (/home/dustin/code/repo/rs_query_builder/sickle-builder)
[snip: all warnings]
    Checking tiberius v0.11.3
    Checking sickle-query v0.1.0 (/home/dustin/code/repo/rs_query_builder/sickle-query)
    Finished dev [unoptimized + debuginfo] target(s) in 2.87s
```

**Comment by dustbort** [comment]
So I wonder if it is the way that rustc is being invoked, which is by rust-analyzer or vscode?

**Comment by dustbort** [comment]
OK, I got the error to show with `cargo build --workspace --all-targets --all-features`. I will open an issue with rust.
